[iOS] [.NET 8] Binding library source generator does not use interface for protocols

Hello,

I'm trying to update my iOS binding library from .NET 7.0 to 8.0, but I believe the code behind source generator is not generating the correct code. When I define a protocol, it's not using the generated interface causing the object to try to inherit from both the NSObject and the protocol class. This results in compiler error CS1721 (cannot have multiple base classes).

This was working correctly in .NET 7.0, could I be missing a breaking change?

### Steps to Reproduce

Create a new binding library and add the following code to the ApiDefinition.cs in a namespace of your choosing.

```cs
using System;
using ObjCRuntime;
using Foundation;
using UIKit;

[Protocol]
[BaseType(typeof(NSObject))]
public interface ReaderProtocol
{
    // @required -(NSString *)firstName;
    [Abstract]
    [Export("firstName")]
    string FirstName { get; }
}

// @interface Reader : NSObject <ReaderProtocol>
[BaseType(typeof(NSObject))]
public interface Reader : ReaderProtocol
{
    // -(id)initWithPreferredReader:(NSString *)name;
    [Export("initWithPreferredReader:")]
    IntPtr Constructor(string name);
}
```
### Expected Behavior

The generated Reader.g.cs file should have a ```Reader``` class with the following signature (net7.0-ios behavior):

```cs
[Register("Reader", true)]
public unsafe partial class Reader : NSObject, IReaderProtocol {
...
}
```

### Actual Behavior

The generated Reader.g.cs file contains a ```Reader``` class with the following signature (net8.0-ios behavior):

```cs
[Register("Reader", true)]
public unsafe partial class Reader : NSObject, ReaderProtocol {
...
}
```

This causes the compiler to throw error CS1721 "Class 'Reader' cannot have multiple base classes: 'NSObject' and 'ReaderProtocol'.

### Environment

<!--
1. How to get version information: https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#version-information
2. Paste into the code block below (between ```)
-->

<details>
<summary>Version information</summary>

```
```

</details>

### Build Logs

<!--
1. How to get a build log: https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#build-logs
2. Place the cursor below this comment block.
3. Drag and drop the build log.
-->

### Example Project (If Possible)

<!--
1. Place cursor below this comment block.
2. Drag and drop the compressed project or files needed to reproduce.
-->

<!--
Switch to the "Preview" tab to ensure your issue renders correctly.
-->

--- a/src/bgen/Generator.cs
+++ b/src/bgen/Generator.cs
@@ -5086,7 +5086,7 @@ public partial class Generator : IMemberGatherer {
 			if (sb.Length > 0)
 				sb.Append (", ");
 			sb.Append (iface.Namespace).Append ('.');
-			if (iface.Assembly.GetName ().Name == "temp")
+			if (IsApiAssembly (iface.Assembly))
 				sb.Append ('I');
 			else if (iface.IsClass)
 				sb.Append ('I');
@@ -5763,7 +5763,7 @@ public partial class Generator : IMemberGatherer {
 				string pname = protocolType.Name;
 				// the extra 'I' is only required for the bindings being built, if it comes from something already
 				// built (e.g. monotouch.dll) then the interface will alreadybe prefixed
-				if (protocolType.Assembly.GetName ().Name == "temp")
+				if (IsApiAssembly (protocolType.Assembly))
 					pname = "I" + pname;
 				var iface = FormatType (type, protocolType.Namespace, pname);
 				if (!implements_list.Contains (iface))
@@ -7262,4 +7262,10 @@ public partial class Generator : IMemberGatherer {
 			return provider?.ToString ();
 		}
 	}
+
+	public bool IsApiAssembly (Assembly assembly)
+	{
+		return assembly == TypeManager.ApiAssembly;
+	}
+
 }

--- a/src/bgen/TypeManager.cs
+++ b/src/bgen/TypeManager.cs
@@ -6,6 +6,8 @@ using System.Reflection;
 #nullable enable
 
 public class TypeManager {
+	public Assembly ApiAssembly { get; }
+
 	Frameworks Frameworks { get; }
 
 	public Type System_Attribute { get; }
@@ -276,6 +278,7 @@ public class TypeManager {
 			throw ErrorHelper.CreateError (4, bindingTouch.CurrentPlatform);
 
 		Frameworks = bindingTouch.Frameworks;
+		ApiAssembly = apiAssembly;
 
 		/* corlib */
 		System_Attribute = Lookup (corlibAssembly, "System", "Attribute");

--- a/tests/generator/BGenTests.cs
+++ b/tests/generator/BGenTests.cs
@@ -8,7 +8,9 @@ using NUnit.Framework;
 using Mono.Cecil;
 using Mono.Cecil.Cil;
 
+using Xamarin;
 using Xamarin.Tests;
+using Xamarin.Utils;
 
 namespace GeneratorTests {
 	[TestFixture ()]
@@ -1431,6 +1433,45 @@ namespace GeneratorTests {
 		{
 			BuildFile (Profile.iOS, "tests/internal-delegate.cs");
 		}
+
+#if NET
+		[Test]
+		public void Issue19612 ()
+		{
+			var profile = Profile.iOS;
+			var filename = Path.Combine (Configuration.SourceRoot, "tests", "generator", "issue19612.cs");
+
+			Configuration.IgnoreIfIgnoredPlatform (profile.AsPlatform ());
+
+			// Compile the temporary assembly and pass the compiled assembly to the generator instead
+			// of relying on the generator to compile.
+			var tmpdir = Cache.CreateTemporaryDirectory ();
+			var tmpassembly = Path.Combine (tmpdir, "temporaryAssembly.dll");
+			var cscArguments = new List<string> ();
+			if (!StringUtils.TryParseArguments (Configuration.DotNetCscCommand, out var cscCommand, out var ex))
+				throw new InvalidOperationException ($"Unable to parse the .NET csc command '{Configuration.DotNetCscCommand}': {ex.Message}");
+			cscArguments.AddRange (cscCommand);
+			var cscExecutable = cscArguments [0];
+			cscArguments.RemoveAt (0);
+			cscArguments.Add (filename);
+			cscArguments.Add ($"/out:{tmpassembly}");
+			cscArguments.Add ("/target:library");
+			cscArguments.Add ($"/r:{Path.Combine (Configuration.DotNetBclDir, "System.Runtime.dll")}");
+			var tf = TargetFramework.Parse (BGenTool.GetTargetFramework (profile));
+			cscArguments.Add ($"/r:{Configuration.GetBindingAttributePath (tf)}");
+			cscArguments.Add ($"/r:{Configuration.GetBaseLibrary (tf)}");
+			var rv = ExecutionHelper.Execute (cscExecutable, cscArguments);
+			Assert.AreEqual (0, rv, "CSC exit code");
+
+			var bgen = new BGenTool ();
+			bgen.Profile = profile;
+			bgen.CompiledApiDefinitionAssembly = tmpassembly;
+			bgen.Defines = BGenTool.GetDefaultDefines (bgen.Profile);
+			bgen.CreateTemporaryBinding (filename);
+			bgen.AssertExecute ("build");
+			bgen.AssertNoWarnings ();
+		}
+#endif
 
 		BGenTool BuildFile (Profile profile, params string [] filenames)
 		{

--- a/tests/generator/BGenTool.cs
+++ b/tests/generator/BGenTool.cs
@@ -36,6 +36,7 @@ namespace Xamarin.Tests {
 		public string? BaseLibrary;
 		public string? AttributeLibrary;
 		public bool ReferenceBclByDefault = true;
+		public string? CompiledApiDefinitionAssembly = null;
 #else
 		public string BaseLibrary = None;
 		public string AttributeLibrary = None;
@@ -133,6 +134,11 @@ namespace Xamarin.Tests {
 			if (CompileCommand.Count > 0) {
 				sb.Add ($"--compile-command");
 				sb.Add (string.Join (" ", StringUtils.QuoteForProcess (CompileCommand.ToArray ())));
+			}
+
+			if (CompiledApiDefinitionAssembly is not null) {
+				sb.Add ($"--compiled-api-definition-assembly");
+				sb.Add (CompiledApiDefinitionAssembly);
 			}
 #endif
 

--- a/tests/generator/issue19612.cs
+++ b/tests/generator/issue19612.cs
@@ -1,0 +1,16 @@
+using System;
+
+using Foundation;
+using ObjCRuntime;
+using UIKit;
+
+namespace iosbindinglib {
+	[Protocol]
+	[BaseType (typeof (NSObject))]
+	public interface ReaderProtocol {
+	}
+
+	[BaseType (typeof (NSObject))]
+	public interface Reader : ReaderProtocol {
+	}
+}

